x/stakeibc/keeper/unbonding_records.go

```diff
@@ -43,171 +69,261 @@ func (k Keeper) CreateEpochUnbondingRecord(ctx sdk.Context, epochNumber uint64)
+//   unbonding has been accounted for
+// Returns the list of messages and the callback data for the ICA
+func (k Keeper) GetUnbondingICAMessages(
+	hostZone types.HostZone,
+	totalUnbondAmount sdkmath.Int,
+	prioritizedUnbondCapacity []ValidatorUnbondCapacity,
+) (msgs []sdk.Msg, unbondings []*types.SplitDelegation, err error) {
+	// Loop through each validator and unbond as much as possible
+	remainingUnbondAmount := totalUnbondAmount
+	for _, validatorCapacity := range prioritizedUnbondCapacity {
+		// Break once all unbonding has been accounted for
+		if remainingUnbondAmount.IsZero() {
+			break
+		}
 
-		// Ignore validators with a zero undelegation amount to prevent a failed transaction on the host
-		if undelegationAmount.IsZero() {
-			continue
+		// Unbond either up to the capacity or up to the total remaining unbond amount
+		// (whichever comes first)
+		var unbondAmount sdkmath.Int
+		if validatorCapacity.Capacity.LT(remainingUnbondAmount) {
+			unbondAmount = validatorCapacity.Capacity
+		} else {
+			unbondAmount = remainingUnbondAmount
 		}
 
-		// Store the ICA transactions
+		remainingUnbondAmount = remainingUnbondAmount.Sub(unbondAmount)
+		unbondToken := sdk.NewCoin(hostZone.HostDenom, unbondAmount)
+
+		// Build the undelegate ICA messages
 		msgs = append(msgs, &stakingtypes.MsgUndelegate{
 			DelegatorAddress: hostZone.DelegationIcaAddress,
-			ValidatorAddress: validatorAddress,
-			Amount:           undelegationAmount,
+			ValidatorAddress: validatorCapacity.ValidatorAddress,
+			Amount:           unbondToken,
 		})
 
-		// Store the split delegations for the callback
-		splitDelegations = append(splitDelegations, &types.SplitDelegation{
-			Validator: validatorAddress,
-			Amount:    undelegationAmount.Amount,
+		// Build the validator splits for the callback
+		unbondings = append(unbondings, &types.SplitDelegation{
+			Validator: validatorCapacity.ValidatorAddress,
+			Amount:    unbondAmount,
 		})
 	}
 
-	// Shouldn't be possible, but if all the validator's had a target unbonding of zero, do not send an ICA
-	if len(msgs) == 0 {
-		return nil, sdkmath.ZeroInt(), nil, nil, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "Target unbonded amount was 0 for each validator")
+	// Sanity check that we had enough capacity to unbond
+	if !remainingUnbondAmount.IsZero() {
+		return msgs, unbondings,
+			fmt.Errorf("unable to unbond full amount (%v) from %v", totalUnbondAmount, hostZone.ChainId)
 	}
 
-	// Store the callback data
-	undelegateCallback := types.UndelegateCallback{
-		HostZoneId:              hostZone.ChainId,
-		SplitDelegations:        splitDelegations,
-		EpochUnbondingRecordIds: epochUnbondingRecordIds,
+	return msgs, unbondings, nil
+}
+
+// Submits undelegation ICA messages for a given host zone
+//
+// The total unbond amount is determined from the epoch unbonding records
+// The unbond amount is then distributed amongst validators proportional to how unbalanced they are.
```

Comment:
This comment line is not 100% true.  We order by how out of balance they are but then fill up to use the least messages.  As written in english you just described the more fair algorithm where we split the amount up and none send their full capacity but we had decided against this since it is more complicated to compute and it often causes way more messages needing to be sent.

Reply:
Poor phrasing on my part, but I was just trying to indicate that validators that are more unbalanced, have more of their delegation targeted. And I was also trying to be more succinct since I described it in detail in other places.

Hows this?
```
// First, the total unbond amount is determined from the epoch unbonding records
// Then that unbond amount is distributed amongst validators by first targeting validators
//   who have a proportionally larger delegation than their entitled to (due to the natural 
//   unbalancing from LSM liquid stakes)
// As a result, unbondings lead to a more balanced distribution of stake across validators
```